program/tests/helpers.rs

```diff
@@ -0,0 +1,1110 @@
+                activation_epoch: u64::MAX, // Bootstrap = instantly effective
+                deactivation_epoch: u64::MAX,
+                voter_pubkey: Pubkey::new_unique(),
+            },
+        );
+
+        tracker
+    }
+
+    /// Track a new stake delegation (called after delegate instruction)
+    pub fn track_delegation(
+        &mut self,
+        stake_pubkey: &Pubkey,
+        stake_amount: u64,
+        activation_epoch: Epoch,
+        voter_pubkey: &Pubkey,
+    ) {
+        self.delegations.insert(
+            *stake_pubkey,
+            TrackedDelegation {
+                stake: stake_amount,
+                activation_epoch,
+                deactivation_epoch: u64::MAX,
+                voter_pubkey: *voter_pubkey,
+            },
+        );
+    }
+
+    /// Mark a stake as deactivating (called after deactivate instruction)
+    pub fn track_deactivation(&mut self, stake_pubkey: &Pubkey, deactivation_epoch: Epoch) {
+        if let Some(delegation) = self.delegations.get_mut(stake_pubkey) {
+            delegation.deactivation_epoch = deactivation_epoch;
+        }
+    }
+
+    /// Calculate aggregate stake history for an epoch (replicates Stakes::activate_epoch)
+    fn calculate_epoch_entry(
+        &self,
+        epoch: Epoch,
+        stake_history: &StakeHistory,
+        new_rate_activation_epoch: Option<Epoch>,
+    ) -> StakeHistoryEntry {
+        self.delegations
+            .values()
+            .map(|tracked| {
+                let delegation = Delegation {
+                    voter_pubkey: tracked.voter_pubkey,
+                    stake: tracked.stake,
+                    activation_epoch: tracked.activation_epoch,
+                    deactivation_epoch: tracked.deactivation_epoch,
+                    ..Delegation::default()
+                };
+
+                delegation.stake_activating_and_deactivating(
+                    epoch,
+                    stake_history,
+                    new_rate_activation_epoch,
+                )
+            })
+            .fold(StakeHistoryEntry::default(), |acc, status| {
+                StakeHistoryEntry {
+                    effective: acc.effective + status.effective,
+                    activating: acc.activating + status.activating,
+                    deactivating: acc.deactivating + status.deactivating,
+                }
+            })
+    }
+}
+
+/// Extension trait that adds stake-aware warping to Mollusk
+pub trait MolluskStakeExt {
+    /// Warp to a slot and automatically update stake history at epoch boundaries
+    ///
+    /// This replicates Banks' behavior from solana-runtime:
+    /// - Bank::warp_from_parent() advances slot
+    /// - Stakes::activate_epoch() aggregates delegations
+    /// - Bank::update_stake_history() writes sysvar
+    fn warp_to_slot_with_stake_tracking(
+        &mut self,
+        tracker: &StakeTracker,
+        target_slot: u64,
+        new_rate_activation_epoch: Option<Epoch>,
+    );
```

Comment:
Given this is emulated behavior, can we write tests or add assertions to give assurance that behavior is 1:1 equivalent?